pyo3-macros-backend/src/module.rs

```diff
@@ -136,23 +127,97 @@ pub fn pymodule_module_impl(mut module: syn::ItemMod) -> Result<TokenStream> {
+                }
+            }
+            Item::Mod(item_mod) => {
+                let is_pymodule_export =
+                    find_and_remove_attribute(&mut item_mod.attrs, "pymodule_export");
+                if is_pymodule_export {
+                    module_items.push(item_mod.ident.clone());
+                    module_items_cfg_attrs.push(get_cfg_attributes(&item_mod.attrs));
+                }
+            }
+            Item::ForeignMod(item) => {
+                if has_attribute(&item.attrs, "pymodule_export") {
+                    bail_spanned!(item.span() => "#[pymodule_export] cannot be used on foreign mods but only on inline mods");
+                }
+            }
+            Item::Trait(item) => {
+                if has_attribute(&item.attrs, "pymodule_export") {
+                    bail_spanned!(item.span() => "#[pymodule_export] cannot be used on traits");
+                }
+            }
+            Item::Const(item) => {
+                if has_attribute(&item.attrs, "pymodule_export") {
+                    bail_spanned!(item.span() => "#[pymodule_export] cannot be used on const");
+                }
+            }
+            Item::Static(item) => {
+                if has_attribute(&item.attrs, "pymodule_export") {
+                    bail_spanned!(item.span() => "#[pymodule_export] cannot be used on statics");
+                }
+            }
+            Item::Macro(item) => {
+                if has_attribute(&item.attrs, "pymodule_export") {
+                    bail_spanned!(item.span() => "#[pymodule_export] cannot be used on macros");
                 }
             }
-            item => {
-                bail_spanned!(item.span() => "only 'use' statements and and pymodule_init functions are allowed in #[pymodule]")
+            Item::ExternCrate(item) => {
+                if has_attribute(&item.attrs, "pymodule_export") {
+                    bail_spanned!(item.span() => "#[pymodule_export] cannot be used on extern crates");
+                }
+            }
+            Item::Impl(item) => {
+                if has_attribute(&item.attrs, "pymodule_export") {
+                    bail_spanned!(item.span() => "#[pymodule_export] cannot be used on impls");
+                }
+            }
+            Item::TraitAlias(item) => {
+                if has_attribute(&item.attrs, "pymodule_export") {
+                    bail_spanned!(item.span() => "#[pymodule_export] cannot be used on trait aliases");
+                }
+            }
+            Item::Type(item) => {
+                if has_attribute(&item.attrs, "pymodule_export") {
+                    bail_spanned!(item.span() => "#[pymodule_export] cannot be used on type aliases");
+                }
+            }
+            Item::Union(item) => {
+                if has_attribute(&item.attrs, "pymodule_export") {
+                    bail_spanned!(item.span() => "#[pymodule_export] cannot be used on unions");
+                }
```

Comment:
All this code only allows nicer error message. Happy to remove it if you prefer

Reply:
I'm in favour of nicer error messages.

Reply:
I guess if we go for the automatic export, then instead we can just change to something like

```
`#[pymodule_export]` may only be used on `use` statements
```

Reply:
Change done!